Accessibility and localization improvements

eduardb opened the pull request:
#### TL;DR

Application was not a11y/l10n compliant, so we made sure it is ^^
#### Problem

Many views were missing a proper content description. Application didn't support RTL locales because it used left/right layout attributes. Also, there was no way to have weather condition strings translated.
#### Solution

Added content descriptions to views where it made sense, extracted translatable strings for weather conditions, and use start/end attributes instead of left/right to better support RTL locales.
##### Test(s) added

 Yes, tested that correct weather description is fetched for various weather condition ids.
##### Screenshots

| Before | After |
| --- | --- |
| ![device-2016-04-19-180118](https://cloud.githubusercontent.com/assets/2426348/14648272/5776fc2a-0659-11e6-98fc-84914679c9db.png) | ![device-2016-04-19-175958](https://cloud.githubusercontent.com/assets/2426348/14648274/5fa1fc2e-0659-11e6-80c8-54d503bd5f92.png) |
| ------ | ----- |
| ![a11y_before](https://cloud.githubusercontent.com/assets/2426348/14751457/ddc9fccc-08c1-11e6-8201-06f8267decfa.gif) | ![a11y_after](https://cloud.githubusercontent.com/assets/2426348/14751449/d0ce35a6-08c1-11e6-9205-dd362cd2862a.gif) |
###### Paired with

no one

Commit:
Address some i18n issues around weather conditions

Commit:
Improve a11y for weather conditions

Commit:
Add RTL support

Commit:
Update gradle

Commit:
Use Resources.getIdentifier() to get weather condition string resource id

Commit:
Move weather condition strings to separate file

Commit:
Update gradle plugin once again

ataulm commented on app/src/main/java/com/example/android/sunshine/app/DetailFragment.java:
```diff
@@ -193,37 +193,46 @@ public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
-            // Read description from cursor and update view
-            String description = data.getString(COL_WEATHER_DESC);
+            // Get description from weather condition ID
+            String description = Utility.getStringForWeatherCondition(getActivity(), weatherId);
```
Does fragment not have getString?

eduardb replied:
It has, but I don't see what difference it makes. There is some logic behind how the weather description is got, and it is extracted in the `Utility` class now.

ataulm commented on app/src/main/java/com/example/android/sunshine/app/DetailFragment.java:
```diff
@@ -193,37 +193,46 @@ public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
             double high = data.getDouble(COL_WEATHER_MAX_TEMP);
             String highString = Utility.formatTemperature(getActivity(), high);
             mHighTempView.setText(highString);
+            mHighTempView.setContentDescription(getString(R.string.a11y_high_temp, highString));
```
:100: setting explicit cd

blundell replied:
@ataulm correct me if I am wrong, but if a view has `setText` then this text would be used for the content description any way?

ataulm replied:
You are correct. Sometimes it's desired to set different CD than the text -
for succinctness or to provide more context.

Generally I would recommend making all these not important for
accessibility and putting a content description on the whole card/row.

On Wed, 20 Apr 2016, 13:46 Paul Blundell, notifications@github.com wrote:

> In app/src/main/java/com/example/android/sunshine/app/DetailFragment.java
> https://github.com/eduardb/Sunshine-Version-2/pull/1#discussion_r60399889
> :
> 
> > ```
> >          // Read high temperature from cursor and update view
> >          boolean isMetric = Utility.isMetric(getActivity());
> > 
> >          double high = data.getDouble(COL_WEATHER_MAX_TEMP);
> >          String highString = Utility.formatTemperature(getActivity(), high);
> >          mHighTempView.setText(highString);
> > ```
> > -            mHighTempView.setContentDescription(getString(R.string.a11y_high_temp, highString));
> 
> @ataulm https://github.com/ataulm correct me if I am wrong, but if a
> view has setText then this text would be used for the content description
> any way?
> 
> —
> You are receiving this because you were mentioned.
> 
> Reply to this email directly or view it on GitHub
> https://github.com/eduardb/Sunshine-Version-2/pull/1/files/eafffa1bb4c0d605c06206d992863949083fe22d#r60399889

blundell commented on app/build.gradle:
```diff
@@ -2,7 +2,7 @@ apply plugin: 'com.android.application'
 android {
     compileSdkVersion 23
-    buildToolsVersion "23.0.2"
+    buildToolsVersion "23.0.3"
```
nice when this is done as a separate PR. Usually on bigger projects updating the tools can make ppl giving your PR a +1 take a whoooole lot longer, so just safer to split for faster merges

eduardb replied:
Makes sense, I didn't think about it because this is a small project.

blundell replied:
yeah exactly

blundell commented on app/src/main/java/com/example/android/sunshine/app/ForecastAdapter.java:
```diff
@@ -75,17 +75,18 @@ public void bindView(View view, Context context, Cursor cursor) {
         ViewHolder viewHolder = (ViewHolder) view.getTag();
 
         int viewType = getItemViewType(cursor.getPosition());
+        int weatherId = cursor.getInt(ForecastFragment.COL_WEATHER_CONDITION_ID); // TODO bad practice!?!?!
```
what is bad practice?

eduardb replied:
Oh snap, it was a comment for myself that I committed by mistake. I think it's bad practice that the adapter has knowledge about the fragment in which is being used.

blundell replied:
more explanatory comments will always help yourself when you come back to the code :-)

eduardb replied:
I resolved the TODO here: d92e389

blundell commented on app/src/main/java/com/example/android/sunshine/app/Utility.java:
```diff
@@ -259,6 +260,36 @@ public static int getArtResourceForWeatherCondition(int weatherId) {
+            stringId = context.getResources().getIdentifier(
+                    String.format(Locale.US, "condition_%d", weatherId),
+                    "string",
+                    context.getPackageName());
```
you could use `BuildConfig.APPLICATION_ID` to avoid the use of context on this line

eduardb replied:
b5b04af

blundell commented on app/src/main/res/values/strings.xml:
```diff
@@ -116,4 +116,63 @@
+  <string name="gentle_breeze">Gentle Breeze</string>
+  <string name="breeze">Breeze</string>
+  <string name="fresh_breeze">Fresh Breeze</string>
+  <string name="strong_breeze">Strong Breeze</string>
```
I would recommend using case sensitive names so that you know exactly what it holds i.e.

`<string name="Strong_Breeze">Strong Breeze</string>`

eduardb replied:
499e714

blundell commented on build.gradle:
```diff
@@ -5,7 +5,7 @@ buildscript {
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:2.1.0-alpha5'
+        classpath 'com.android.tools.build:gradle:2.1.0-beta3'
```
👍 living on the edge

eduardb replied:
Always 😎

xrigau replied:
beta is less edgy than alpha #justSayin

ataulm replied:
hahaha

Commit:
Move weather projection from ForecastFragment to WeatherContract

This way, the ForecastAdapter won't have to know that it's used by the ForecastFragment

Commit:
Use constant BuildConfig.APPLICATION_ID instead of context.getPackageName()

Commit:
Use case sensitive names for string resources

... so we know exactly what they hold

xrigau commented on app/src/main/AndroidManifest.xml:
```diff
@@ -16,6 +16,7 @@
     android:allowBackup="true"
     android:icon="@mipmap/ic_launcher"
     android:label="@string/app_name"
+    android:supportsRtl="true"
```
👍

xrigau commented on app/src/main/java/com/example/android/sunshine/app/DetailFragment.java:
```diff
@@ -193,37 +193,46 @@ public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
             // Read humidity from cursor and update view
             float humidity = data.getFloat(COL_WEATHER_HUMIDITY);
             mHumidityView.setText(getActivity().getString(R.string.format_humidity, humidity));
+            mHumidityView.setContentDescription(mHumidityView.getText());
```
isn't this what paul is talking about? this is redundant (same for the 2 below) cc @ataulm

ataulm replied:
yep this one is redundant as TalkBack would read the TextView's text property anyway.

eduardb replied:
7429b7b

xrigau commented on app/src/main/java/com/example/android/sunshine/app/Utility.java:
```diff
@@ -259,6 +260,36 @@ public static int getArtResourceForWeatherCondition(int weatherId) {
+            if (stringId == 0) {
+                return context.getString(R.string.condition_unknown, weatherId);
+            }
+            return context.getString(stringId);
```
unnecessary return

eduardb replied:
85d1bdf

xrigau commented on app/src/main/java/com/example/android/sunshine/app/Utility.java:
```diff
@@ -259,6 +260,36 @@ public static int getArtResourceForWeatherCondition(int weatherId) {
+                    "string",
+                    BuildConfig.APPLICATION_ID);
+
+            if (stringId == 0) {
```
sounds like an if-else `if string == 0 do this otherwise do ->` https://github.com/eduardb/Sunshine-Version-2/pull/1/files#diff-7537c0d5330829deab3c57d11a53aa5dR280

xrigau replied:
ah sorry I misread that, but this `if` could be moved to line 290 (as in - doesn't need to be a nested `if`)

ataulm replied:
nested `ifs` can generally be extracted into smaller methods so you have at most one level of indentation per method

eduardb replied:
cd2bb74

xrigau commented on app/src/main/java/com/example/android/sunshine/app/Utility.java:
```diff
@@ -259,6 +260,36 @@ public static int getArtResourceForWeatherCondition(int weatherId) {
+     * @param weatherId from OpenWeatherMap API response
+     * @return string for the weather condition. null if no relation is found.
+     */
+    public static String getStringForWeatherCondition(Context context, int weatherId) {
```
this could be unit tested 👅  I see 4 unit tests 💥

eduardb replied:
Is this what you meant? 3af1eb0

xrigau commented:
good stuff! just a couple of comments but the rest LGTM 👍

Commit:
Remove unnecessary return in Utility.getStringForWeatherCondition()

ataulm commented on app/src/main/res/values/strings.xml:
```diff
@@ -116,4 +116,63 @@
   <string name="empty_forecast_list_invalid_location">No weather information available. The location in settings is not recognized by the weather server.</string>
 
+  <!-- A11y -->
+  <string name="a11y_forecast">Forecast: <xliff:g id="condition">%1$s</xliff:g></string>
```
sik xliffin

ataulm replied:
the xliff tag also takes an `example` attribute I think, so you could provide an example of what the token's value to help translators understand the context.

eduardb replied:
31ebfe4

ataulm commented on app/src/main/java/com/example/android/sunshine/app/ForecastAdapter.java:
```diff
@@ -75,44 +77,46 @@ public void bindView(View view, Context context, Cursor cursor) {
-        boolean isMetric = Utility.isMetric(context);
+        // For accessibility, we don't want a content description for the icon field
+        // because the information is repeated in the description view and the icon
+        // is not individually selectable
```
I thought I saw a string resource for the icon's content description. if it's not used, should be deleted.

Note, ImageView's do not need content descriptions - the lint check is wrong (or out of date at least)!

An ImageView will be ignored by TalkBack unless you put an explicit content description, add `importantForAccessibility="yes"` or make it focusable/clickable.

ImageButtons on the other hand are buttons and as such are clickable by default, so you need a content description or it needs to be marked as `importantForAccessibiilty="no"` to be ignored. If it's a button that does something and you hide it from TalkBack, then an affordance for TalkBack users must be available somewhere else.

cc @blundell

blundell replied:
what versions of Android do you refer to though?

ataulm replied:
16+ at least but I guess all.

ataulm commented:
👍 lgtm, but :x: until you add a gif that shows the effect this had on TalkBack. You can "display text" in TalkBack's developer settings (found at the bottom of TalkBack Settings)

😁

Commit:
Add tests for `Utility.getStringForWeatherCondition()`

Commit:
Move nested `if` outside

Commit:
Add `example` attribute to `xliff` tags

Commit:
Remove redundant set CD on TextViews

eduardb commented:
@ataulm I added the gifs, the "display text" is very useful, thanks for the tip 💯

ataulm commented:
👍 lgetm

you can use the DOWN THEN RIGHT gesture to open the global context menu, and then pressing and holding in the center, drag up to "Read from top".

This'll go through all the elements from the top of your screen (it won't scroll if you have a list though), so you can see the order in which it iterates through your views.

eduardb commented:
Another day, another thing that I learn from you :D
Thanks guys for the review, I really appreciate it @ataulm @blundell @xrigau

xrigau commented:
![](https://media.giphy.com/media/XreQmk7ETCak0/giphy.gif)
